File: src/main/java/com/company/spacetrans/entity/Planet.java
package com.company.spacetrans.entity;

import io.jmix.core.entity.annotation.JmixGeneratedValue;
import io.jmix.core.metamodel.annotation.Composition;
import io.jmix.core.metamodel.annotation.JmixEntity;

import javax.persistence.*;
import java.util.UUID;

@JmixEntity
@Table(name = "PLANET")
@Entity
public class Planet extends AstronomicalBody{
    @JmixGeneratedValue
    @Column(name = "ID", nullable = false)
    @Id
    private UUID id;

    @Column(name = "SEMI_MAJOR_AXIS")
    private Double semiMajorAxis;

    @Column(name = "ORBITAL_PERIOD")
    private Double orbitalPeriod;

    @Column(name = "ROTATION_PERIOD")
    private Double rotationPeriod;

    @JoinColumn(name = "ATMOSPHERE_ID")
    @Composition
    @OneToOne(fetch = FetchType.LAZY)
    private Atmosphere atmosphere;

    @Column(name = "RINGS")
    private Boolean rings;

    public Boolean getRings() {
        return rings;
    }

    public void setRings(Boolean rings) {
        this.rings = rings;
    }

    public Double getRotationPeriod() {
        return rotationPeriod;
    }

    public void setRotationPeriod(Double rotationPeriod) {
        this.rotationPeriod = rotationPeriod;
    }

    public Double getOrbitalPeriod() {
        return orbitalPeriod;
    }

    public void setOrbitalPeriod(Double orbitalPeriod) {
        this.orbitalPeriod = orbitalPeriod;
    }

    public Double getSemiMajorAxis() {
        return semiMajorAxis;
    }

    public void setSemiMajorAxis(Double semiMajorAxis) {
        this.semiMajorAxis = semiMajorAxis;
    }

    public Atmosphere getAtmosphere() {
        return atmosphere;
    }

    public void setAtmosphere(Atmosphere atmosphere) {
        this.atmosphere = atmosphere;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }
}

File: frontend/src/jmix/entities/Planet.ts
import { AstronomicalBody } from "./AstronomicalBody";
import { Atmosphere } from "./Atmosphere";
export class Planet extends AstronomicalBody {
  static NAME = "Planet";
  id?: string;
  semiMajorAxis?: any | null;
  orbitalPeriod?: any | null;
  rotationPeriod?: any | null;
  atmosphere?: Atmosphere | null;
  rings?: boolean | null;
}
export type PlanetViewName =
  | "_base"
  | "_instance_name"
  | "_local"
  | "planet-fetch-plan";
export type PlanetView<V extends PlanetViewName> = V extends "_base"
  ? Pick<
      Planet,
      | "id"
      | "name"
      | "semiMajorAxis"
      | "orbitalPeriod"
      | "rotationPeriod"
      | "rings"
      | "mass"
      | "picture"
    >
  : V extends "_instance_name"
  ? Pick<Planet, "id" | "name">
  : V extends "_local"
  ? Pick<
      Planet,
      | "id"
      | "semiMajorAxis"
      | "orbitalPeriod"
      | "rotationPeriod"
      | "rings"
      | "name"
      | "mass"
      | "picture"
    >
  : V extends "planet-fetch-plan"
  ? Pick<
      Planet,
      | "id"
      | "name"
      | "semiMajorAxis"
      | "orbitalPeriod"
      | "rotationPeriod"
      | "rings"
      | "mass"
      | "picture"
    >
  : never;
